[LLDB] Add optional callback function to `TypeMatcher`

Nerixyz opened the pull request:
To be able to have multiple formatters that match on the same type name but for different types, this adds a callback that can be specified, which is called with the match candidate. As [mentlerd pointed out](https://discourse.llvm.org/t/formatting-msvc-stl-types/86667/12?u=nerixyz), it's similar to passing `--recognizer-function`. This isn't implemented as another match type, because the `TypeMatcher` still needs to have _some_ name for [`GetMatchString`](https://github.com/llvm/llvm-project/blob/f1575de4c5de9268f92eea1641af755a477e4ee4/lldb/include/lldb/DataFormatters/FormattersContainer.h#L120-L126).

The only use I currently see here is for the C++ STL, specifically libstdc++ and MSVC's STL.
In https://github.com/llvm/llvm-project/pull/143177, I used this to add formatters for `std::string` and friends from MSVC's STL.

One potential issue of adding the callback here is that the summary/synthetic will be cached for a type name - disregarding the callback. The scenario here would be two shared libraries that use a different type with the same name and two formatters (e.g. they use two different standard libraries). Once the first type was formatted, its formatter is cached. Now, the second type will always be formatted with the wrong formatter (the cached one), even though the callback for the cached one wouldn't match.
For one, `--recognizer-function` has the same issue and secondly, I don't think this is really realistic.

Nerixyz committed:
[LLDB] Add optional callback function to `TypeMatcher`

jimingham commented:
You could even make the cache hold the "match string" and a LIST of matching formatters.  If a VO passes the typename string match, then we run through the list of matching formatters and choose the first one whose validator accepts this VO.

That way we could have both the STL version formatters in the cache but never choose the wrong one when dealing with heterogeneous targets.

jimingham commented:
Another way to do this would be to make the data formatters held per target not per debugger.  But that's a much bigger change.

Nerixyz commented:
> When you are about to use a cached formatter, just run the validator function from that formatter on the candidate VO. If it doesn't pass the validator, fall back to the full search.

I think for this, it's easier to save the validator per formatter/summary/synthetic, because we need the validator when retrieving from the cache, right? That shouldn't be a big issue, just a refactoring.

> You could even make the cache hold the "match string" and a LIST of matching formatters. If a VO passes the typename string match, then we run through the list of matching formatters and choose the first one whose validator accepts this VO.

I think in almost all cases, having one entry in the cache is better. Sure, there might be scenarios where you have different types with the same name, but realistically, when you're formatting you're formatting (e.g. at a breakpoint), you'll most likely have one type per name.

Nerixyz commented:
It's still missing tests, but I updated it to validate in the formatter. I also updated https://github.com/llvm/llvm-project/pull/143177 where it's implemented for std::string.

jimingham commented:
> > When you are about to use a cached formatter, just run the validator function from that formatter on the candidate VO. If it doesn't pass the validator, fall back to the full search.
> 
> I think for this, it's easier to save the validator per formatter/summary/synthetic, because we need the validator when retrieving from the cache, right? That shouldn't be a big issue, just a refactoring.
> 
> > You could even make the cache hold the "match string" and a LIST of matching formatters. If a VO passes the typename string match, then we run through the list of matching formatters and choose the first one whose validator accepts this VO.
> 
> I think in almost all cases, having one entry in the cache is better. Sure, there might be scenarios where you have different types with the same name, but realistically, when you're formatting you're formatting (e.g. at a breakpoint), you'll most likely have one type per name.

If you have two Targets in lldb, and they each have a variant of some type that you've discriminated by a TypeValidator, then it's quite possible that Target A and Target B hit breakpoints in a location that contains a local of the given type in sequence.  For instance, if I were trying to follow some RPC from client to server and back, I'd be stopping with the "client" view of the message first, then stopping on receipt with the "server" view of the message, etc.

So one  {typename, validator} pair will get set in the cache when you first stop in the client.  Then when you stop in the server, we'll answer that the type is not cached because the validator fails.  I think we'll then cache the {typename, second_validator} so at that point we're getting no benefit from caching...

jimingham commented:
Also, when you go to write tests be sure to write a test with this case.  If there were a way to add the validator from Python you could do this pretty easily.  Maybe you can do it a unit test w/o going that far?

Nerixyz commented:
> I think we'll then cache the {typename, second_validator} so at that point we're getting no benefit from caching...

Yes, if you're going back and forth, then the current caching won't help.

I'm not that familiar with all the use cases of LLDB. In my thinking, you'd break in the client, evaluate multiple variables and then break in the server where you'd also evaluate multiple variables. My assumption would be that some of these have the same type. In that case you'd get two misses but all other requests would be cached.

Zooming out a bit more: what's the concrete use-case where this would be a problem? As explained above, having two types with the same name but different layouts seems really rare. Moreover, you'd need to be debugging both at the same time to run into this. And even then, it's not a logic bug but a performance bug.

> Maybe you can do it a unit test w/o going that far?

I'll try to do this.

labath commented:
A low-tech way to support both STL versions would be to do a dispatch inside the data formatter:
```
StdStringSummaryProvider(value) {
  if (IsGnuSTL(value))
    return GnuStringSummaryProvider(value);
  return MSVCStringSummaryProvider(value);
}
```

It's not a complete solution as it doesn't allow for a distributed setup (vendor A writes a formatter for type `Foo`, vendor B writes a formatter for its type `Foo`, user C needs to debug both `Foo` types, so it installs both formatters), but for formatters that are a part of lldb (i.e. STL types), that should work fine. And maybe STL types are the only place where this matters?

Nerixyz commented:
> I've described a use case a couple times already. If you have a client-server architecture, it would be really nice to be able to debug both the client and the server in the same lldb process, because then you can add tooling on top of lldb to do things like "step in across remote protocol dispatch". And in client-server systems it is often the case that the client and the server run on different OS-es. Jim

Sorry if I sounded mean before, that wasn't my intention at all. I assumed you meant just using lldb-server, but with this explanation it makes sense why you'd debug both and now I can see why you run into this. 

What do you think about caching at most two instances per kind (summary/snythetic/formatter) per entry? The two isn't from "client&server" but from having at most two different ones (gnu&msvc).

Nerixyz commented:
> A low-tech way to support both STL versions would be to do a dispatch inside the data formatter

For summaries this would work, but it Gerts more involved for synthetic children providers, where you'd need to check for the kind in every method.

Though now that I think about it, couldn't I inspect the type before creating the frontend for a synthetic provider and create the appropriate frontend?

labath commented:
> Though now that I think about it, couldn't I inspect the type before creating the frontend for a synthetic provider and create the appropriate frontend?

That is what I was imagining, yes.

Nerixyz commented:
Thank you for the feedback!

Closing this, as it's not needed to allow different summary/synthetic providers for one type name.
